src/server/http/prototype_pollution.test.ts

```diff
@@ -0,0 +1,123 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { IncomingMessage } from 'http';
+import { setupPrototypePollutionValidation } from './prototype_pollution';
+
+// we need to grab a reference to the function that is added
+// with server.ext('onPostAuth', fn) so we can emulate the payload
+const getHandler = () => {
+  const mockServer = {
+    ext: jest.fn(),
+  };
+  setupPrototypePollutionValidation(mockServer as any);
+  expect(mockServer.ext).toHaveBeenCalledWith('onPostAuth', expect.anything());
+  return mockServer.ext.mock.calls[0][1];
+};
+
+const createMockH = () => {
+  return {
+    continue: Symbol(),
+  };
+};
+
+const expectSuccess = (obj: any) => {
+  const handler = getHandler();
+  const mockH = createMockH();
+  const result = handler(obj, mockH);
+  expect(result).toBe(mockH.continue);
+};
+
+const expectFailure = (obj: any) => {
+  const handler = getHandler();
+  const mockH = createMockH();
+  expect(() => handler(obj, mockH)).toThrowErrorMatchingSnapshot();
+};
+
+// to ensure we're ignoring incoming messages, we're creating a circular dependency
+// and ensuring that we're still getting a success
+test(`ignores IncomingMessage`, () => {
+  const message = new IncomingMessage(null as any);
+  message.socket = message as any;
+  expectSuccess({
+    payload: message,
+  });
+});
+
+test(`fails on circular references`, () => {
```

Comment:
I'm considering changing the logic to no longer fail on circular references, and to instead support them. If we've already checked the object and all of it's children, there's no reason why we necessarily have to check it again.

Reply:
I'm not necessarily opposed, but I liked the idea of failing on circular references. If we come across a circular reference, chances are it's a scenario we haven't thought about yet, so we wouldn't have considered the implications of such a payload.

Reply:
Yeah, same here. I was seeing behavior elsewhere and jumped to a conclusion. I need to adjust some logic when dealing with the circular logic detection and Arrays, since they are objects, but if we have two arrays with the same "primitive" keys, they're being detected as a cycle.

Reply:
I can't replicate the issue that I was seeing on Friday with Arrays, so I'm going to leave this logic in place.